Testing.Rmd

```diff
@@ -442,159 +769,1106 @@ It's hard to give good general advice about writing tests, but you might find th
-  expect_floor_equal("week",   "2009-08-02 00:00:00")
-  expect_floor_equal("month",  "2009-08-01 00:00:00")
-  expect_floor_equal("year",   "2009-01-01 00:00:00")
+#### Built-in `skip()` functions
+
+Similar to testthat's built-in expectations, there is a family of `skip()` functions that anticipate some common situations.
+These functions often relieve you of the need to write a custom skipper.
+Here are some examples of the most useful `skip()` functions:
+
+```{r eval = FALSE}
+test_that("foo api returns bar when given baz", {
+  skip_if(api_unavailable(), "API not available")
+  ...
+})
+test_that("foo api returns bar when given baz", {
+  skip_if_not(api_available(), "API not available")
+  ...
+})
+
+skip_if_not_installed("sp")
+skip_if_not_installed("stringi", "1.2.2")
+
+skip_if_offline()
+skip_on_cran()
+skip_on_os("windows")
+```
+
+#### Dangers of skipping
+
+One challenge with skips is that they are currently completely invisible in CI — if you automatically skip too many tests, it's easy to fool yourself that all your tests are passing when in fact they're just being skipped!
+In an ideal world, your CI/CD would make it easy to see how many tests are being skipped and how that changes over time.
+
+*2022-06-01: Recent changes to GitHub Actions mean that we will likely have better test reporting before the second edition of this book is published. Stay tuned!*
+
+It's a good practice to regularly dig into the `R CMD check` results, especially on CI, and make sure the skips are as you expect.
+But this tends to be something you have to learn through experience.
+
+### Mocking
+
+The practice known as mocking is when we replace something that's complicated or unreliable or out of our control with something simpler, that's fully within our control.
+Usually we are mocking an external service, such as a REST API, or a function that reports something about session state, such as whether the session is interactive.
+
+The classic application of mocking is in the context of a package that wraps an external API.
+In order to test your functions, technically you need to make a live call to that API to get a response, which you then process.
+But what if that API requires authentication or what if it's somewhat flaky and has occasional downtime?
+It can be more productive to just *pretend* to call the API but, instead, to test the code under your control by processing a pre-recorded response from the actual API.
+
+Our main advice about mocking is to avoid it if you can.
+This is not an indictment of mocking, but just a realistic assessment that mocking introduces new complexity that is not always justified by the payoffs.
+
+Since most R packages do not need full-fledged mocking, we do not cover it here.
+Instead we'll point you to the packages that represent the state-of-the-art for mocking in R today:
+
+* mockery: <https://github.com/r-lib/mockery>
+* mockr: <https://krlmlr.github.io/mockr/>
+* httptest: <https://enpiar.com/r/httptest/>
```

Comment:
more plugging! demo of httptest https://books.ropensci.org/http-testing/httptest.html

you should add httptest2 as well, and maybe webfakes even if it is not mocking stricto sensu